### ISSUE TITLE
Share more code with xml5ever

### ISSUE THREAD

**Issue by SimonSapin** [servo/html5ever#266]
Yes, after discussion in https://github.com/Ygg01/xml5ever/issues/30 we’ve moved the xml5ever code into this repository with the intention to share more code. However much of that has yet to happen. Rough plan:

* Make every crate in the repo use a single [workspace](http://doc.crates.io/manifest.html#the-workspace-section)
* Make sure Travis-CI is running every test
* Rename the the `html5ever_atoms` crate to `markup5ever` and update html5ever and xml5ever to use it. This is a breaking change, so increment version numbers accordingly.
* Make it so that users of either html5ever or xml5ever don’t need to have an explicit dependency to markup5ever (which should be an implementation detail). One way to do this is to have `pub use markup5ever::*;` at the root of each crate.
* Find things that exist in both html5ever and xml5ever, and move them to markup5ever. Judge on a case-by-case basis when there are differences.

  For example `TokenSink` has some methods in common, but both parsers have some methods unique to it in their respective `TokenSink`. Since this trait is implemented by users of the libraries, the shared `TokenSink` should have the union of all methods (to support both parsers) and provide default implementations of methods where that makes sense (to reduce the overhead for users that only care about one parser).

  The `Parser` and `BytesParser` types in `html5ever::driver` are busted for handling `</script>`, we might want to remove them or redesign them. (Don’t bother adding it to xhtml5ever.)

**Comment by Ygg01** [comment]
Sounds like a plan. I am assuming following structure for workspace:

```
├── Cargo.toml
├── markupl5ever
    ├── Cargo.toml
    └── src
    └──...
├── html5ever
    ├── Cargo.toml
    └── src
    └──...
├── xml5ever
    ├── Cargo.toml
    └── src
    └──...

```

**Comment by Ygg01** [comment]
Here is the WIP: https://github.com/Ygg01/html5ever/tree/xhtml5ever

Questions I have arise are following:

1. What does [`travis-build.sh`](https://github.com/Ygg01/html5ever/blob/master/scripts/travis-build.sh) does? It's called in `.travis.yml`, but seems to essentially, just call `cargo test`? Or is the capture part doing something interesting?

2. What QualName to use? Apparently, there were discussions around QualName having a local, prefix and namespace, but there were fears about it exploding size? Could I abstract over it like so:

```
struct QualName<T>{
    pub localname: Localname,
    pub namespace: Namespace,
    pub prefix: T,
}

type XmlQualName = QualName<Prefix>;
type HtmlQualName = QualName<()>;
```
3. This one is more of a question for me, but I think there isn't a huge value in EmptyTag (e.g. `<tagName/>`) and Short tag (e.g. `</>`). Without those two, I could reuse htmltags. However, I might not change this now, because changes will be numerous and hard even without it; and this change requires changing of XML parsing.

**Comment by SimonSapin** [comment]
1. `travis-build.sh` could be inlined in `.travis.yml`. Having a separate file avoid doubt of what escaping is required to nest bash syntax inside YAML string syntax.

   Rust’s test harness has a feature where the output of a test is "captured" and shown only that test fails, together with the failure message. Since tests can run in parallel on multiple threads, this avoids interleaved output and shows which output comes from which test. This feature is based on `#[unstable]` APIs from the standard library.

   html5ever uses [a fork of the `test` crate](https://crates.io/crates/rustc-test) that exposes APIs that are `#[unstable]` in the standard library copy, in order to generate the list of tests dynamically. To make this fork usable on stable Rust, any feature that uses `#[unstable]` std APIs are optional and disabled by default. `--features "rustc-test/capture"` enables the capture feature.
2. I think we should not worry about the size of `QualName`. Servo only uses it transiently during parsing. What is stored in `Element` for example looks like this:

   ```rust
   pub struct Element {
       // …
       local_name: LocalName,
       namespace: Namespace,
       prefix: Option<DOMString>,
       // …
   }
   ```

   Let’s have a single `QualName` type that contains `Option<Prefix>`. This will often be `None` in html5ever, except in `TreeBuilderActions::adjust_foreign_attributes`. [*Adjust foreign attributes* in the spec](https://html.spec.whatwg.org/multipage/#adjust-foreign-attributes) can set a non-null prefix.

3. Empty tags are also part of XML 1. Even if they weren’t, this alone doesn’t feel like a strong enough reason to drop features from XML 5. And this is the interface between tokenizers and their respective tree builders, which are not interchangeable and typically implementation details not used directly by users of these libraries. Keeping two `TagKind` enum sounds fine.

4. Git does not explicitly store information about renames. Instead it can reconstruct that information when querying history. For `git log`, this is enabled by the `--follow` (for a single file) or `-M` a.k.a. `--find-renames` (for diffs) options. This is based on searching for new files that look similar to files removed in the same commit. If the files are not identical, there’s a heuristic to decide whether they’re close enough. So if you’re both renaming and making significant changes to a file, consider doing so in separate commits to help this heuristic.

**Comment by karan1276** [comment]
Hello, I am a student. I recently started contributing on servo. I found my self doing only "easy" bugs, which gets boring after some time. I would like to work on servo and actually understand whats going on. To begin with i have decided to work on html5ever first and understand how parsing is done at lower levels, If you guys could sort of mentor me that will be great( @SimonSapin  @Ygg01 )

Looks like this issue has lots of stuff to do, may i be of any help?

**Comment by SimonSapin** [comment]
I think @Ygg01 has already started working on this, so you should coordinate together :)

**Comment by Ygg01** [comment]
@karan1276 I will gladly accept a PR to my PR :P So what do you want to do? Want do you want to know about? You can look at xml5ever as a simplified version of html5ever. Basically, xml5ever and hmtl5ever tokenization is similar, but tree building is waaaay different, because HTML has insertion modes.

@SimonSapin Could I partially commit this? I think I did [items 1-4](https://github.com/servo/html5ever/issues/266#issue-221518400) on that list, but recently some PRs broke my commits :( Last bullet point is really numerous, and might take more sense to divide.

**Comment by SimonSapin** [comment]
@Ygg01 one thing is that if there are breaking API changes I’d prefer to batch them as much as reasonable. Do you think there’s work that can be deferred to later without introducing more breakage then? In any case feel free to send PRs :)

**Comment by Ygg01** [comment]
@SimonSapin Ok, I think I got most of stuff done. Only TokenSink and TreeSink are left.

That last might take a bit, since xml5ever `TokenSink` bitrotted a bit. E.g. it still has old  [`fn process_token`](https://github.com/servo/html5ever/commit/7a268f3c66ef680323e39b6ade07875b9595479c#diff-09238cea356947c5f9b11475004c0888L109).

My thoughts are keep those methods but mark them for deprecated, and/or remove them in next version.

**Comment by SimonSapin** [comment]
Don’t spend effort trying to unify `TokenSink`. The XML5 tokenizer only needs to work with the XML5 tree builder, and the HTML tokenizer only needs to work with the HTML tree builder. They are not interchangeable.

Regarding deprecation, we’re making breaking changes anyway so I wouldn’t worry too much about it.

`pub use markup5ever::*;` sounds good. No risk the we forget to re-export something.
